public/pages/AnomalyCharts/utils/anomalyChartUtils.ts

```diff
@@ -138,3 +150,281 @@ export const disabledHistoryAnnotations = (
     },
   ];
 };
+
+export const ANOMALY_HEATMAP_COLORSCALE = [
+  [0, '#F2F2F2'],
+  [0.2, '#F2F2F2'],
+  [0.2, '#F7E0B8'],
+  [0.4, '#F7E0B8'],
+  [0.4, '#F2C596'],
+  [0.6, '#F2C596'],
+  [0.6, '#ECA976'],
+  [0.8, '#ECA976'],
+  [0.8, '#E78D5B'],
+  [1, '#E8664C'],
+];
+
+export enum AnomalyHeatmapSortType {
+  SEVERITY = 'by_severity',
+  OCCURRENCES = 'by_occurrences',
+}
+
+const getColorForValue = (value: number) => {
+  if (
+    value >=
+    ANOMALY_HEATMAP_COLORSCALE[ANOMALY_HEATMAP_COLORSCALE.length - 1][0]
+  ) {
+    return ANOMALY_HEATMAP_COLORSCALE[ANOMALY_HEATMAP_COLORSCALE.length - 1][1];
+  }
+
+  if (value <= ANOMALY_HEATMAP_COLORSCALE[0][0]) {
+    return ANOMALY_HEATMAP_COLORSCALE[0][1];
+  }
+
+  for (let i = 0; i < ANOMALY_HEATMAP_COLORSCALE.length - 1; i++) {
+    if (
+      value >= ANOMALY_HEATMAP_COLORSCALE[i][0] &&
+      value < ANOMALY_HEATMAP_COLORSCALE[i + 1][0]
+    ) {
+      return ANOMALY_HEATMAP_COLORSCALE[i + 1][1];
+    }
+  }
+};
+
+const NUM_CELLS = 20;
+
+export const HEATMAP_X_AXIS_DATE_FORMAT = 'MM-DD HH:mm YYYY';
+
+const buildBlankStringWithLength = (length: number) => {
+  let result = '';
+  for (let i = 0; i < length; i++) {
+    result += ' ';
+  }
+  return result;
+};
+
+export const getAnomaliesHeatmapData = (
+  anomalies: any[],
+  dateRange: DateRange,
+  sortType: AnomalyHeatmapSortType = AnomalyHeatmapSortType.SEVERITY,
+  displayTopNum: number
+): PlotData[] => {
+  const entityAnomaliesMap = getEntityAnomaliesMap(anomalies);
+  if (isEmpty(entityAnomaliesMap)) {
+    // put placeholder data so that heatmap won't look empty
+    for (let i = 0; i < displayTopNum; i++) {
+      const blankStrValue = buildBlankStringWithLength(i);
+      entityAnomaliesMap.set(blankStrValue, []);
```

Comment:
why put various length empty strings in the map? is it so a new mapping is created for each entry? is this for showing a default empty heatmap in the case of no anomalies?

Reply:
yeah. It is just to show default empty heatmap in the case of no anomalies. kind of hacky. If I only put 10 same length whitespace string as entity value, aka Y values, it will show like this:
![Screen Shot 2020-10-16 at 5 57 58 PM](https://user-images.githubusercontent.com/59710443/96324935-29e31580-0fd9-11eb-8217-2534435ba007.png)

Reply:
yeah, makes sense. could you maybe add a comment describing why you had to do this though?

Reply:
sure. added comment.